From 897ae38e65dad2b8e3e20310dba345cc677d413e Mon Sep 17 00:00:00 2001
From: lbarapat <lbarapat@qti.qualcomm.com>
Date: Fri, 14 Aug 2026 12:25:37 +0530
Subject: [PATCH] paho-mqtt-c: add ssl_external_sign_cb to
 MQTTClient_SSLOptions

Adds three new fields to MQTTClient_SSLOptions (struct_version 6):
  - ssl_external_sign_cb: callback for TA-backed CertificateVerify
  - ssl_external_signing_algorithm: TLS signature algorithm
  - ssl_external_signing_method: enable flag for external signing path

Also adds SSL_THREAD_IN_PROGRESS connect state to Clients.h to track
the ssl_handshake_thread lifecycle and prevent use-after-free on destroy.

These fields allow the TLS private key to remain in QNX TrustZone;
OpenSSL calls the callback for CertificateVerify instead of accessing
the key directly.

Signed-off-by: lbarapat <lbarapat@qti.qualcomm.com>
---
 src/Clients.h    |   8 ++-
 src/MQTTClient.h | 118 +++++++++++++++++++++++++++++++++++++++--------
 2 files changed, 106 insertions(+), 19 deletions(-)

diff --git a/src/Clients.h b/src/Clients.h
--- a/src/Clients.h
+++ b/src/Clients.h
@@ -104,6 +104,10 @@
 /** SSL connection in progress */
 #define SSL_IN_PROGRESS  0x2
 /** Websocket connection in progress */
+/** SSL handshake being completed by a dedicated thread (external signing path).
+ *  The background thread sets this state after spawning ssl_handshake_thread. */
+#define SSL_THREAD_IN_PROGRESS  0x6
+/** Websocket connection in progress */
 #define WEBSOCKET_IN_PROGRESS   0x3
 /** TCP completed, waiting for MQTT ACK */
 #define WAIT_FOR_CONNACK 0x4
@@ -114,8 +118,9 @@
 
 /**
  * Data related to one client
+ * The entire structure is initialized to 0 on creation, so all fields default to 0.
  */
-typedef struct
+typedef struct Clients
 {
 	char* clientID;					      /**< the string id of the client */
 	const char* username;					/**< MQTT v3.1 user name */
@@ -132,6 +137,7 @@
 	networkHandles net;             /**< network info for this client */
 	int msgID;                      /**< the MQTT message id */
 	int keepAliveInterval;          /**< the MQTT keep alive interval */
+	int savedKeepAliveInterval;     /**< saved keep alive interval, in case reset by server keep alive */
 	int retryInterval;              /**< the MQTT retry interval for QoS > 0 */
 	int maxInflightMessages;        /**< the max number of inflight outbound messages we allow */
 	willMessages* will;             /**< the MQTT will message, if any */

diff --git a/src/MQTTClient.h b/src/MQTTClient.h
index 6fd31f6..ae428a2 100644
--- a/src/MQTTClient.h
+++ b/src/MQTTClient.h
@@ -40,11 +40,13 @@
  *
  * @endcond
  * @cond MQTTClient_main
- * @mainpage MQTT Client library for C
+ * @mainpage MQTT Client library for C (MQTTClient)
  * &copy; Copyright 2009, 2022 IBM Corp., Ian Craggs and others
  *
  * @brief An MQTT client library in C.
  *
+ * Version 1.3.14
+ *
  * These pages describe the original more synchronous API which might be
  * considered easier to use.  Some of the calls will block.  For the new
  * totally asynchronous API where no calls block, which is especially suitable
@@ -97,6 +99,7 @@
  * <li>@ref wildcard</li>
  * <li>@ref qos</li>
  * <li>@ref tracing</li>
+ * <li>@ref HTTP_proxies</li>
  * </ul>
  * @endcond
  */
@@ -112,6 +115,7 @@
 #endif
 
 #include <stdio.h>
+#include <stdint.h>
 /*
 /// @endcond
 */
@@ -123,6 +127,8 @@
 #include "MQTTSubscribeOpts.h"
 #if !defined(NO_PERSISTENCE)
 #include "MQTTClientPersistence.h"
+#else
+#define MQTTCLIENT_PERSISTENCE_NONE 1
 #endif
 
 /**
@@ -179,9 +185,13 @@
   */
  #define MQTTCLIENT_BAD_MQTT_VERSION -11
 /**
- * Return code: protocol prefix in serverURI should be tcp://, ssl://, ws:// or wss://
- * The TLS enabled prefixes (ssl, wss) are only valid if a TLS version of the library
- * is linked with.
+ * Return code: protocol prefix in serverURI should be:
+ * @li @em tcp:// or @em mqtt:// - Insecure TCP
+ * @li @em ssl:// or @em mqtts:// - Encrypted SSL/TLS
+ * @li @em ws:// - Insecure websockets
+ * @li @em wss:// - Secure web sockets
+ * The TLS enabled prefixes (ssl, mqtts, wss) are only valid if a TLS
+ * version of the library is linked with.
  */
 #define MQTTCLIENT_BAD_PROTOCOL -14
  /**
@@ -401,7 +411,8 @@ typedef void MQTTClient_connectionLost(void* context, char* cause);
 /**
  * This function sets the callback functions for a specific client.
  * If your client application doesn't use a particular callback, set the
- * relevant parameter to NULL. Calling MQTTClient_setCallbacks() puts the
+ * relevant parameter to NULL (except for message arrived, which must be given).
+ * Calling MQTTClient_setCallbacks() puts the
  * client into multi-threaded mode. Any necessary message acknowledgements and
  * status communications are handled in the background without any intervention
  * from the client application. See @ref async for more information.
@@ -494,13 +505,21 @@ LIBMQTT_API int MQTTClient_setPublished(MQTTClient handle, void* context, MQTTCl
  * this function.
  * @param serverURI A null-terminated string specifying the server to
  * which the client will connect. It takes the form <i>protocol://host:port</i>.
- * Currently, <i>protocol</i> must be <i>tcp</i>, <i>ssl</i>, <i>ws</i> or <i>wss</i>.
- * The TLS enabled prefixes (ssl, wss) are only valid if a TLS version of the library
- * is linked with.
- * For <i>host</i>, you can
- * specify either an IP address or a host name. For instance, to connect to
- * a server running on the local machines with the default MQTT port, specify
- * <i>tcp://localhost:1883</i>.
+ * Currently, <i>protocol</i> must be:
+ * <br>
+ * @em tcp:// or @em mqtt:// - Insecure TCP
+ * <br>
+ * @em ssl:// or @em mqtts:// - Encrypted SSL/TLS
+ * <br>
+ * @em ws:// - Insecure websockets
+ * <br>
+ * @em wss:// - Secure web sockets
+ * <br>
+ * The TLS enabled prefixes (ssl, mqtts, wss) are only valid if a TLS
+ * version of the library is linked with.
+ * For <i>host</i>, you can specify either an IP address or a host name. For
+ * instance, to connect to a server running on the local machines with the
+ * default MQTT port, specify <i>tcp://localhost:1883</i>.
  * @param clientId The client identifier passed to the server when the
  * client connects to it. It is a null-terminated UTF-8 encoded string.
  * @param persistence_type The type of persistence to be used by the client:
@@ -661,12 +680,14 @@ typedef struct
 	/** The eyecatcher for this structure.  Must be MQTS */
 	char struct_id[4];
 
-	/** The version number of this structure. Must be 0, 1, 2, 3, 4 or 5.
+	/** The version number of this structure. Must be 0, 1, 2, 3, 4, 5 or 6.
 	 * 0 means no sslVersion
 	 * 1 means no verify, CApath
 	 * 2 means no ssl_error_context, ssl_error_cb
 	 * 3 means no ssl_psk_cb, ssl_psk_context, disableDefaultTrustStore
 	 * 4 means no protos, protos_len
+	 * 5 means no ssl_external_signing_method, ssl_external_sign_cb,
+	 *            ssl_external_signing_algorithm
 	 */
 	int struct_version;
 
@@ -765,9 +786,30 @@ typedef struct
 	 * Exists only if struct_version >= 5
 	 */
 	unsigned int protos_len;
+
+	/** Enable external signing for mTLS CertificateVerify.
+	 * Set to 1 to use ssl_external_sign_cb; ssl_opts.privateKey is ignored.
+	 * Exists only if struct_version >= 6. */
+	int ssl_external_signing_method;
+
+	/** Signing callback: sign(algorithm, data, data_len, out, max_out, out_len)
+	 * data is the pre-hashed digest; out receives the DER-encoded signature.
+	 * Return non-zero on success. Exists only if struct_version >= 6. */
+	unsigned int (*ssl_external_sign_cb)(uint16_t signature_algorithm,
+	                                     const void *data, size_t data_len,
+	                                     void *out, size_t max_out,
+	                                     size_t *out_len);
+
+	/** TLS SignatureScheme for ssl_external_sign_cb.
+	 * Use SSL_SIGN_ECDSA_SECP256R1_SHA256 (0x0403) for EC P-256.
+	 * Exists only if struct_version >= 6. */
+	uint16_t ssl_external_signing_algorithm;
 } MQTTClient_SSLOptions;
 
-#define MQTTClient_SSLOptions_initializer { {'M', 'Q', 'T', 'S'}, 5, NULL, NULL, NULL, NULL, NULL, 1, MQTT_SSL_VERSION_DEFAULT, 0, NULL, NULL, NULL, NULL, NULL, 0, NULL, 0 }
+/** TLS SignatureScheme: ECDSA with secp256r1 and SHA-256 (RFC 8446 s4.2.3) */
+#define SSL_SIGN_ECDSA_SECP256R1_SHA256  0x0403
+
+#define MQTTClient_SSLOptions_initializer { {'M', 'Q', 'T', 'S'}, 6, NULL, NULL, NULL, NULL, NULL, 1, MQTT_SSL_VERSION_DEFAULT, 0, NULL, NULL, NULL, NULL, NULL, 0, NULL, 0, 0, NULL, 0 }
 
 /**
   * MQTTClient_libraryInfo is used to store details relating to the currently used
@@ -802,6 +844,12 @@ LIBMQTT_API MQTTClient_nameValue* MQTTClient_getVersionInfo(void);
  * values to 0 (NULL for pointers). A #keepAliveInterval setting of 0 prevents
  * correct operation of the client and so you <b>must</b> at least set a value
  * for #keepAliveInterval.
+ *
+ * Suitable default values are set in the following initializers:
+ * - MQTTClient_connectOptions_initializer: for MQTT 3.1.1 non-WebSockets
+ * - MQTTClient_connectOptions_initializer5: for MQTT 5.0 non-WebSockets
+ * - MQTTClient_connectOptions_initializer_ws: for MQTT 3.1.1 WebSockets
+ * - MQTTClient_connectOptions_initializer5_ws: for MQTT 5.0 WebSockets
  */
 typedef struct
 {
@@ -951,24 +999,34 @@ typedef struct
 	 */
 	const MQTTClient_nameValue* httpHeaders;
 	/**
-	 * HTTP proxy
-	 */
+	* The string value of the HTTP proxy. Examples:
+	*  - http://your.proxy.server:8080/
+	*  - http://user:pass@my.proxy.server:8080/
+	*/
 	const char* httpProxy;
 	/**
-	 * HTTPS proxy
-	 */
+	* HTTPS proxy setting. See ::MQTTClient_connectOptions.httpProxy and the section @ref HTTP_proxies.
+	*/
 	const char* httpsProxy;
 } MQTTClient_connectOptions;
 
+/** Initializer for connect options for MQTT 3.1.1 non-WebSocket connections */
 #define MQTTClient_connectOptions_initializer { {'M', 'Q', 'T', 'C'}, 8, 60, 1, 1, NULL, NULL, NULL, 30, 0, NULL,\
 0, NULL, MQTTVERSION_DEFAULT, {NULL, 0, 0}, {0, NULL}, -1, 0, NULL, NULL, NULL}
 
+/** Initializer for connect options for MQTT 5.0 non-WebSocket connections */
 #define MQTTClient_connectOptions_initializer5 { {'M', 'Q', 'T', 'C'}, 8, 60, 0, 1, NULL, NULL, NULL, 30, 0, NULL,\
 0, NULL, MQTTVERSION_5, {NULL, 0, 0}, {0, NULL}, -1, 1, NULL, NULL, NULL}
 
+/** Initializer for connect options for MQTT 3.1.1 WebSockets connections.
+  * The keepalive interval is set to 45 seconds to avoid webserver 60 second inactivity timeouts.
+  */
 #define MQTTClient_connectOptions_initializer_ws { {'M', 'Q', 'T', 'C'}, 8, 45, 1, 1, NULL, NULL, NULL, 30, 0, NULL,\
 0, NULL, MQTTVERSION_DEFAULT, {NULL, 0, 0}, {0, NULL}, -1, 0, NULL, NULL, NULL}
 
+/** Initializer for connect options for MQTT 5.0 WebSockets connections.
+  * The keepalive interval is set to 45 seconds to avoid webserver 60 second inactivity timeouts.
+  */
 #define MQTTClient_connectOptions_initializer5_ws { {'M', 'Q', 'T', 'C'}, 8, 45, 0, 1, NULL, NULL, NULL, 30, 0, NULL,\
 0, NULL, MQTTVERSION_5, {NULL, 0, 0}, {0, NULL}, -1, 1, NULL, NULL, NULL}
 
@@ -1375,6 +1433,14 @@ LIBMQTT_API void MQTTClient_freeMessage(MQTTClient_message** msg);
   */
 LIBMQTT_API void MQTTClient_free(void* ptr);
 
+/**
+  * This function is used to allocate memory to be used or freed by the MQTT C client library,
+  * especially the data in user persistence. This is needed on Windows when the client library
+  * and application program have been compiled with different versions of the C compiler.
+  * @param size The size of the memory to be allocated.
+  */
+LIBMQTT_API void* MQTTClient_malloc(size_t size);
+
 /**
   * This function frees the memory allocated to an MQTT client (see
   * MQTTClient_create()). It should be called when the client is no longer
@@ -1943,4 +2009,20 @@ exit:
     20130528 163909.209 Heap scan end
   * @endcode
   * @endcond
+  *
+* @page HTTP_proxies HTTP Proxies
+  * The use of HTTP proxies can be controlled by environment variables or API calls.
+  *
+  * The ::MQTTClient_connectOptions.httpProxy and ::MQTTClient_connectOptions.httpsProxy fields
+  * of the ::MQTTClient_connectOptions structure override any settings in the environment.
+  *
+  * If the environment variable PAHO_C_CLIENT_USE_HTTP_PROXY is set to TRUE, then the
+  * http_proxy or https_proxy (lower case only) environment variables are used, for plain
+  * TCP and TLS-secured connections respectively.
+  *
+  * The no_proxy environment variable can be used to exclude certain hosts from using an
+  * environment variable chosen proxy. This does not apply to a proxy selected through the API.
+  * The no_proxy environment variable is lower case only, and is a list of comma-separated
+  * hostname:port values. Suffixes are matched (e.g. example.com will match test.example.com).
+  *
   */
-- 
2.34.1

